**Code reviewer**

¡Hola en esta nueva aventura!

Mi nombre es **Ivan Hernandez** y seré tu revisor de código para este proyecto.

Revisaré todo el cuaderno con preguntas o comentarios dejados atrás. Te animo a responder los comentarios con cualquier duda o solución a la recomendación o problema presentado **(con la celda en color azul)**, esa es la mejor manera de entender y aprender en este nuevo mundo de datos y programación. Seguiré la misma lista de verificación y los mismos pasos en los criterios de evaluación del proyecto.

 Importante, <span style="color:red">por favor no cambies mis comentarios</span>, es una buena forma de seguir tu aprendizaje por mí y otros revisores de código y finalmente, pongo un ejemplo del color del código y su significado.

<div class="alert alert-success">
<h1> Comentarios del revisor </h1>
Excelente trabajo y ejecución
</div>

<div class="alert alert-warning">
<h1> Comentarios del revisor </h1>
Comentario no crítico; si hay menos de tres de tales comentarios, el proyecto puede aceptarse sin su refinamiento.</div>

<div class="alert alert-danger">
<h1> Comentarios del revisor </h1>
Error que debe corregirse, de lo contrario el proyecto no será aceptado.
</div>

<div class="alert alert-info">
<h1> Comentarios del estudiante</h1>
Comentario de ejemplo
</div>

# Analisis Empresarial

No estoy satisfecho con el hecho de que muchos de nuestros clientes estén usando planes heredados en la compañía móvil Megaline. Me gustaría desarrollar un modelo que pueda analizar el comportamiento de nuestros clientes y recomendar uno de los nuevos planes de Megaline: Smart o Ultra.

Para esta tarea de clasificación, debo crear un modelo que pueda elegir el plan adecuado. Ya he completado el paso de procesamiento de datos, así que puedo empezar directamente a crear el modelo.

## Importacion de librerias

Vamos a cargar las librerias y separar los datos dependiendo de nuestro objetivo para los conjuntos de entrenamiento, validacion y prueba.

In [1]:
#Importamos todas las librerias que nos puedan ser utiles
import pandas as pd
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression

## Visualizacion y segmentacion de datos

### Visualizacion

In [2]:
#Cargamos el dataframe en una variable
df = pd.read_csv('/datasets/users_behavior.csv')

In [3]:
#Visualizamos informacion del dataframe
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


In [4]:
#Una descripcion del dataframe
df.describe()

calls      minutes     messages       mb_used     is_ultra
count  3214.000000  3214.000000  3214.000000   3214.000000  3214.000000
mean     63.038892   438.208787    38.281269  17207.673836     0.306472
std      33.236368   234.569872    36.148326   7570.968246     0.461100
min       0.000000     0.000000     0.000000      0.000000     0.000000
25%      40.000000   274.575000     9.000000  12491.902500     0.000000
50%      62.000000   430.600000    30.000000  16943.235000     0.000000
75%      82.000000   571.927500    57.000000  21424.700000     1.000000
max     244.000000  1632.060000   224.000000  49745.730000     1.000000

In [5]:
#Una muestra del dataframe
df

calls  minutes  messages   mb_used  is_ultra
0      40.0   311.90      83.0  19915.42         0
1      85.0   516.75      56.0  22696.96         0
2      77.0   467.66      86.0  21060.45         0
3     106.0   745.53      81.0   8437.39         1
4      66.0   418.74       1.0  14502.75         0
...     ...      ...       ...       ...       ...
3209  122.0   910.98      20.0  35124.90         1
3210   25.0   190.36       0.0   3275.61         0
3211   97.0   634.44      70.0  13974.06         0
3212   64.0   462.32      90.0  31239.78         0
3213   80.0   566.09       6.0  29480.52         1

[3214 rows x 5 columns]

Vemos que el conjunto de datos no tiene datos problematicos, pues ya ha pasado por el preprocesamiento de datos. Eso significa que podemos ir directo a nuestro analisis.

### Segmentacion de datos

In [6]:
#Separemos los datos de entrenamiento y validacion
df_train, df_valid = train_test_split(df, test_size=0.25, random_state=54321)

#Separemos el grupo de testeo


#Hagamos los features y target
features_train = df_train.drop(['is_ultra'], axis=1)
target_train = df_train['is_ultra']
features_valid = df_valid.drop(['is_ultra'], axis=1)
target_valid = df_valid['is_ultra']

<div class="alert alert-success">
<h1> Comentarios del revisor </h1>
Muy buen trabajo!
</div>

## Prueba de modelos

Vamos a comenzar por probar los modelos que puedan ayudarnos mas, es decir, los que puedan clasificar como son: DecisionTreeClassifier y RandomForestClassifier.

### DecisionTreeClassifier

In [7]:
best_model = 0
best_result = 0
score_1 = 0
for depth in range(1, 50):
    model_1 = DecisionTreeClassifier(random_state=54321, max_depth=depth) # crea un modelo con la profundidad proporcionada
    model_1.fit(features_train, target_train) # entrena el modelo 
    score_1 = model_1.score(features_valid, target_valid)
    if score_1 > best_model:
        best_model = score_1
        best_result = depth
        
print("Exactitud del mejor modelo en el conjunto de validacion (depth= {}): {}".format(best_result, best_model))

Exactitud del mejor modelo en el conjunto de validacion (depth= 8): 0.7947761194029851


In [8]:
x_train, x_test, y_train, y_test = train_test_split(df.drop('is_ultra', axis=1), df['is_ultra'], test_size=0.25, random_state=54321)

model_dtc = DecisionTreeClassifier(random_state=54321, max_depth=8)
model_dtc.fit(x_train, y_train)

predictions_1 = model_dtc.predict(x_test)
accuracy_1 = accuracy_score(y_test, predictions_1)
print("Exactitud en el conjunto de prueba:", accuracy_1)

Exactitud en el conjunto de prueba: 0.7947761194029851


Podemos ver que en el conjunto de validacion y prueba, el modelo supera el umbral de exactitud requerido, por lo cual se puede tomar en consideracion. Veamos como iran los otros para considerar los dos resultados mas altos.

### RandomForestClassifier

In [9]:
best_score = 0
best_est = 0
for est in range(1, 51, 10):
    model_2 = RandomForestClassifier(random_state=54321, n_estimators=est, max_depth = 20)
    model_2.fit(features_train, target_train)
    score_2 = model_2.score(features_valid, target_valid)
    if score_2 > best_score:
        best_score = score_2
        best_est = est
        
print("Accuracy del mejor modelo en el conjunto de validación (n_estimators = {}): {}".format(best_est, best_score))


Accuracy del mejor modelo en el conjunto de validación (n_estimators = 41): 0.8022388059701493


In [10]:
model_rfc = RandomForestClassifier(random_state = 54321, n_estimators = 21, max_depth= 16)
model_rfc.fit(x_train, y_train)

predictions_2 = model_rfc.predict(x_test)
accuracy_2 = accuracy_score(y_test, predictions_2)
print("Exactitud en el conjunto de prueba:", accuracy_2)

Exactitud en el conjunto de prueba: 0.8009950248756219


Este bosque inclusive tiene un resultado mas alto, por lo que se puede considerar como el mejor resultado hasta ahora. Despues de intentar con algunas profundidades, su mejor resultado fue con un valor de 16 y 21 numero de estimados.

### LogisticRegression

In [11]:
model_3 = LogisticRegression(random_state=54321, solver='liblinear')
model_3.fit(features_train, target_train)
score_train = model_3.score(features_train, target_train)
score_valid = model_3.score(features_valid, target_valid)

print("Accuracy del modelo de regresión logística en el conjunto de entrenamiento:", score_train)
print("Accuracy del modelo de regresión logística en el conjunto de validación:", score_valid)

Accuracy del modelo de regresión logística en el conjunto de entrenamiento: 0.7178423236514523
Accuracy del modelo de regresión logística en el conjunto de validación: 0.6940298507462687


Vemos que los resultados de la regresion lineal no supera o iguala el umbral de exactitud, por lo cual se descarta.

### DecisionTreeRegressor

In [12]:
best_model_2 = None
best_result_2 = 10000
best_depth = 0

for depth in range(1, 6):
    model_4 = DecisionTreeRegressor(max_depth=depth, random_state=54321)
    model_4.fit(features_train, target_train)
    predictions_valid_1 = model_4.predict(features_valid)
    result_1 = mean_squared_error(target_valid, predictions_valid_1) ** 0.5
    if result_1 < best_result_2:
        best_model_2 = model_4
        best_result_2 = result_1
        best_depth = depth

print(f"RMSE del mejor modelo en el conjunto de validación (max_depth = {best_depth}): {best_result_2}")

RMSE del mejor modelo en el conjunto de validación (max_depth = 5): 0.40033089897845153


Vemos que la exactitud de el arbol de regresion no es alta y tampoco iguala al umbral requerido, teniendo un error cuadratico de 0.4.

### RandomForestRegressor

In [13]:
best_model_3 = None
best_result_3 = 10000
best_depth_2 = 0
best_est_2 = 0
for est in range(1, 51, 10):
    for depth in range(1, 11):
        model_5 = RandomForestRegressor(random_state = 54321, n_estimators = est, max_depth = depth)
        model_5.fit(features_train, target_train)
        predictions_valid_2 = model_5.predict(features_valid)
        result_2 = mean_squared_error(target_valid, predictions_valid_2)
        if result_2 < best_result_3:
            best_model_3 = model_5
            best_result_3 = result_2
            best_est_2 = est
            best_depth_2 = depth
            
print("RECM del mejor modelo en el conjunto de validación", best_result_3, "n_estimators:", best_est_2, "best_depth:", depth)

RECM del mejor modelo en el conjunto de validación 0.14749667144498332 n_estimators: 31 best_depth: 10


Podemos ver que el bosque de regresion si pudo igualar el umbral, pero no iguala a la exactitud del arbol clasificador o bosque clasificador, por lo cual no es nuestra primera opcion.

<div class="alert alert-danger">
<h1> Comentarios del revisor </h1>
    
Muy buen trabajo, pero no es necesario crear por cada modelo un conjunto de train y test diferentes, de hecho dado que les pasas el mismo seed. Están quedando exactamente iguales al del principio. 
    
</div>

<div class="alert alert-info">
<h1> Comentarios del estudiante</h1>
Listo!
</div>

<div class="alert alert-success">
<h1> Comentarios del revisor V2</h1>
Muy buen trabajo!
</div>

## Uso del mejor modelo

Nuestro mejor modelo fue el bosque clasificador, por lo que vamos a usarlo con el dataset ya que supero el conjunto de prueba y validacion.

In [15]:
final_features = df.drop(['is_ultra'], axis=1)
final_target = df['is_ultra']

final_features_train, final_features_valid, final_target_train, final_target_valid = train_test_split(final_features, final_target, test_size=0.25, random_state=54321)
final_features_valid, final_features_test, final_target_valid, final_target_test = train_test_split(final_features_valid, final_target_valid, test_size=0.5, random_state=54321)

# Entrenamos el mejor modelo
best_final_model = RandomForestClassifier(random_state=54321, n_estimators=21, max_depth=16)
best_final_model.fit(final_features_train, final_target_train)

# Hacemos las predicciones a partir del mejor modelo
final_predictions_valid = best_final_model.predict(final_features_valid)

# Calculamos la precision en el conjunto de validacion
final_accuracy = accuracy_score(final_target_valid, final_predictions_valid)
print(f"Precision del mejor modelo en el conjunto de validacion: {final_accuracy}")


Precision del mejor modelo en el conjunto de validacion: 0.8059701492537313


Podemos ver que la precision del modelo final supera el umbral de exactitud, y despues de bastantes pruebas se puede concluir que es el modelo mas apto para nuestro analisis que solicitan.

## Conclusion Final

Finalmente podemos ver que depende mucho, pero el modelo mas apto segun la investigacion es el bosque clasificador con un resultado de: 0.8059701492537313.

Tambien se debe considerar si es que queremos mas exactitud o rapidez en cuanto a nuestro trabajo. Sabiendo que el bosque no es el modelo mas apto para velocidad porque busca entre varias opciones de arboles, pero tiene una precision mas alta se tomo en cuenta como el mas optimo.

Si hubieramos querido velocidad pudieramos haber usado la regresion lineal en caso de haber superado el umbral de exactitud, o el arbol clasificador que tambien obtuvo un valor alto.

Se hizo los procesos de revisar que no haya hiperajustes o subajustes, por lo que la certitud del trabajo no es obsoleta.

<div class="alert alert-success">
<h1> Comentarios del revisor </h1>
    
Muy buen trabajo con la implementación de los modelos y todo el proyecto, solo hay un pequeño tema asociado a los conjuntos de entrenamiento que no será problema arreglar. 
    
Saludos!
    
</div>